keps/sig-storage/20190120-ExecutionHook-apichange.md

```diff
@@ -0,0 +1,146 @@
+
+This proposal is aimed to address that limitation by providing an `ExecutionHook`
+in the `Container` struct. The snapshot controller will look up this hook before
+taking a snapshot and execute it accordingly.
+
+## Motivation
+
+The volume snapshot feature allows creating/deleting volume snapshots, and the
+ability to create new volumes from a snapshot natively using the Kubernetes API.
+
+However, application consistency is not guaranteed. An user has to figure out how
+to quiece an application before taking a snapshot and unquiece it after taking
+the snapshot.
+
+So we want to introduce an `ExecutionHook` to facilitate the quiesce and unquiesce
+actions when taking a snapshot. There is an existing lifecycle hook in the `Container`
+struct. The lifecycle hook is called immediately after a container is created or
+immediately before a container is terminated. The proposed execution hook is not
+tied to the start or termination time of the container. Instead it is called before
+or after taking a snapshot while the container is still running.
+
+### Goals
+
+`ExecutionHook` is introduced to define actions that can be taken on a container.
+Specifically it can be used to perform a quiese operation before a volume snapshot
+is taken and then perform an unquiesce operation after a volume snapshot is
+taken to resume the application.
+
+## Non-Goals
+
+This proposal does not provide exact command included in the `ExecutionHook`
+because every application has a different requirement.
+
+## Proposal
+
+An `ExecutionHook` is defined in the following.
+
+```
+// ExecutionHook defines a specific action that should be taken with retries and timeouts
+type ExecutionHook struct {
+        // Command to execute for a particular trigger
+        Handler `json:",inline" protobuf:"bytes,1,opt,name=handler"`
+        // How long the controller should try/retry to execute the hook before giving up
+        RetryTimeOutSeconds int64 `json:"retryTimeOutSeconds,omitempty" protobuf:"varint,2,opt,name=retryTimeOutSeconds"`
+        // Number of retries
+        NumRetries int64 `json:"numRetries,omitempty" protobuf:"varint,3,opt,name=numRetries"`
+}
+```
+
+A `QuiesceUnquiesceHook` struct contains a `Quiesce` and `Unquiesce` `ExecutionHook`.
+
+```
+// QuiesceUnquiesceHook includes a Quiesce action and an Unquiesce action
+type QuiesceUnquiesceHook struct {
+        // The hook to quiesce the application
+        Quiesce *ExecutionHook `json:"quiesce,omitempty" protobuf:"bytes,1,opt,name=quiesce"`
+        // The hook to unquiesce the application
+        Unquiesce *ExecutionHook `json:"unquiesce,omitempty" protobuf:"bytes,2,opt,name=unquiesce"`
+}
+```
+
+A `QuiesceUnquiesceHook` is added to the `Container` struct.
+
+```
+// A single application container that you want to run within a pod.
+type Container struct {
+	......
+        // Defines the hook to quiesce and unquiesce an application
```

Comment:
We need to talk about delivery and define the guarantees.

Who delivers this?  The controller or kubelet?

If the storage controller:  How is it authenticated (e.g. it can't be arbitrarily open to anyone)?  How do they actually trigger an exec?

If the kubelet: How do we tell kubelet to go do it?  What is the trigger from the storage controller?

What guarantees do we make about delivery?  Do we guarantee it gets delivered at all?  Do we guarantee it gets delivered exactly once?  Do we guarantee it gets delivered at least once?

How does the controller know it is done?  Does the app indicate somehow that it is complete, or is this expected to be synchronous?

How do we access-control this?

We need to have answers to all these (and more, I am sure) and they need to be documented.  See https://github.com/kubernetes/kubernetes/issues/24957  for more exploration of this idea.

Reply:
Sure.  We need to provide a lot more details in the spec.  Will take a look of kubernetes/kubernetes#24957 for ideas.

Reply:
queiscing  an application in order to achieve application consistent snapshots using block level snapshots is tricky. The general outline of what you probably want to do looks like this

1. Call the queisce hook to flush any memory and/or WALs  to the filesystem, sync the filesystem storage media, freeze the file system, and block and application level writes.
2. Take the block level snapshot.
3. Call unqueisce to unfreeze the filesystem and allow application level writes.

I think the lifecycle of all the operations is tied.

The hook in (1) needs to return success or failure. If it fails the system should not do (2). If (1) succeeds, without respect to (2), the system must do (3). If (1) succeeds, again without respect to (2), and (3) fails, you've probably broken the application, and you need a way to deal with it. If (2) fails at any point in time, you still need to do (3).

I think it follows that (1) needs to be implemented as idempotent with at least once delivery guarantees. If you do (2) without (1) you are at best crash consistent (and probably not even that).

(3) needs to be idempotent as well.

I think you need to describe how the system will function in more detail in order to motivate the API addition and for people to consider the API in the context of its proposed implementation.

Reply:
Addressed comments in the updated spec.